apply_frame_axis0 applies function to first group twice

javadnoorb opened the pull request:
- [x] closes #21609
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

jorisvandenbossche committed:
DOC: update vendored numpydoc version

Numpydoc upstream at 21a194e9b42ebe95a70475e35cb69bcb4801ff36

jorisvandenbossche committed:
DOC: use numpydoc_use_blockquotes=False for compatibility

jorisvandenbossche committed:
DOC: change to numpydoc: add option to use member listing for attributes

javadnoorb committed:
removed colorbars from _handle_shared_axes when called by scatterplot and hexbin

javadnoorb committed:
removed colorbars from _handle_shared_axes when called by scatterplot and hexbin

javadnoorb committed:
added a debug global variable

l736x committed:
DOC: Improve the docstrings of CategoricalIndex.map (#20286)

shelvinskyi committed:
DOC: update the DataFrame.to_hdf() docstirng (#20186)

jorisvandenbossche committed:
DOC: enable docstring on DataFrame.columns/index (#20385)

jorisvandenbossche committed:
DOC: general docstring formatting fixes (#20449)

minggli committed:
parameterize tests in scalar/timedelta (#20428)

javadnoorb committed:
removed whitespaces

masongallo committed:
API & BUG: allow list-like y argument to df.plot & fix integer arg to x,y (#20000)

* Add support for list-like y argument

* update whatsnew

* add doc change for y

* Add test cases and fix position args

* don't copy save cols ahead of time and update whatsnew

* address fdbck

WillAyd committed:
Removed dead groupby code (#20457)

minggli committed:
EHN: allow zip compression in `to_pickle`, `to_json`, `to_csv` (#20394)

TomAugspurger committed:
ENH: Sorting of ExtensionArrays (#19957)

igorcadelima committed:
DOC: Improve the docstring of DataFrame.transpose() (#20254)

* DOC: Improve the docstring of DataFrame.transpose()

Co-authored-by: Carlos Eduardo Moreira dos Santos <cems@cemshost.com.br>

thismakessand committed:
DOC: Update pandas.Series.copy docstring (#20261)

4kxz committed:
DOC: update the pandas.DataFrame.plot.box docstring (#20373)

Imanflow committed:
DOC: make deprecation warning more visible with red box (#20357)

jreback committed:
TST: clean deprecation warnings & some parametrizing (#20467)

TomAugspurger committed:
REF: Mock all S3 Tests (#20409)

* REF: Mock all S3 Tests

Closes https://github.com/pandas-dev/pandas/issues/19825

minggli committed:
DOC: docstring to series.unique (#20474)

TomAugspurger committed:
TST: Fixed version comparison (#20469)

* TST: Fixed version comparison

This failed to skip for 3.5.x because the micro component made it False.

* Use pandas.compat

* More pandas compat

jreback committed:
TST: 32-bit compat for categorical factorization tests (#20482)

reidy-p committed:
API: Preserve int columns in to_dict('index') (#20444)

tv3141 committed:
DOC: Fix broken dependency links (#20471)

* Change bottleneck link to github repository page.

An issue exists for the broken documentation link on pypy https://github.com/kwgoodman/bottleneck/issues/173

* Fix link to fastparquet documentation.

dcreekp committed:
DOC: update the pandas.Series.str.startswith docstring (#20458)

* DOC: update the pandas.Series.str.startswith docstring

* DOC: update the pandas.Series.str.startswith docstring

* DOC: update the pandas.Series.str.startswith docstring 2

* DOC: update the pandas.Series.str.startswith docstring 3

* DOC: update the pandas.Series.str.startswith docstring 4

JQGoh committed:
BUG: dropna() on single column timezone-aware values (#13407) (#20422)

jreback committed:
TST: test_nanops some parametrize & catch warnings (RuntimeWarning: All-Nan slice in tests) (#20484)

* TST: test_nanops some parametrize & catch warnings (RuntimeWarning: All-NaN slice in tests)

* COMPAT: work around deprecation warning on non-equal dtype comparisons

closes #20011

* WARN: bincount minlength deprecation warning

ibrahimsharaf committed:
ENH: DataFrame.pivot accepts a list of values (#18636)

jorisvandenbossche committed:
BUG: raise error when setting cached properties (#20487)

samuelsinayoko committed:
DOC: update the DataFrame.stack docstring (#20430)

TomAugspurger committed:
CI: Fixed deprecationWarning (#20489)

Closes #20479

Cheukting committed:
DOC: update the isna, isnull, notna and notnull docstring (#20459)

jorisvandenbossche committed:
CLN: remove deprecated infer_dst keyword (#20490)

TomAugspurger committed:
Parametrized NA sentinel for factorize (#20473)

GGordonGordon committed:
DEPR: deprecate get_ftype_counts (GH18243) (#20404)

fdroessler committed:
DOC: update the Series.str.join docstring (#20463)

jorisvandenbossche committed:
DOC: change to numpydoc: use isdatadescriptor instead of isgetsetdescriptor to check attributes

This is needed to ensure .columns / .index / cached properties are recognized as an attribute

TomAugspurger committed:
ENH/API: ExtensionArray.factorize (#20361)

TomAugspurger committed:
DOC: add guide on shared docstrings (#20016)

cbrnr committed:
Set pd.options.display.max_columns=0 by default (#17023)

Change `max_columns` to `0` (automatically adapt the number of displayed columns to the actual terminal width)

jreback committed:
DOC: whatsnew edits

TomAugspurger committed:
ENH: Support ExtensionArray in Groupby (#20502)

dcreekp committed:
DOC: update the pandas.Series.str.endswith docstring (#20491)

Hamishpk committed:
DOC: Improving the docstring of Series.str.upper and related (#20462)

mrw34 committed:
DOC: update the Series.between docstring (#20443)

javadnoorb committed:
Update _core.py

removing PATH_MODE debugging flag

javadnoorb committed:
Update _core.py

remove whitespaces

mmngreco committed:
CLN: Removed not necessary bn switch decorator on nansum (#20481)

TomAugspurger committed:
DOC: Fix various warnings (#20509)

javadnoorb committed:
Update _tools.py

javadnoorb committed:
included tests for scatterplot and hexbin plot to ensure colorbar does not exclude x-axis label and tick-labels

javadnoorb committed:
Update test_frame.py

remove whitespace

jreback committed:
DOC: whatsnew edits

javadnoorb committed:
Update test_frame.py

changing for PEP8

javadnoorb committed:
Update _tools.py

changing for PEP8

javadnoorb committed:
fixing assert message

javadnoorb committed:
Update test_frame.py

fixing lint E231

kylebarron committed:
DOC: Add comparison with Stata page to documentation (#19945)

* Add Comparison with Stata to documentation

* Add comparison_with_stata.rst to index.rst.template

* Fix section underlining

* Small fixes

pulkitmaloo committed:
DOC: Update missing_data.rst (#20424)

javadnoorb committed:
fixing style issues

javadnoorb committed:
Update v0.23.0.txt

plotting bugfixes

javadnoorb committed:
Merge branch 'master' into master

javadnoorb committed:
Update test_frame.py

extra comments pointing to issues

javadnoorb committed:
remove whitespace

javadnoorb committed:
Inline backend (#2)

* minor change in comment

* added condition on inline backend

* added condition on inline backend

* issue with floats determining the position of axes corners in ipython

javadnoorb committed:
Update _tools.py

javadnoorb committed:
Update _core.py

javadnoorb committed:
Update _core.py

javadnoorb committed:
Update v0.23.0.txt

javadnoorb committed:
Update _core.py

javadnoorb committed:
Inline backend - fixing colorbar axis position roundoff error locally

* minor change in comment

* added condition on inline backend

* added condition on inline backend

* issue with floats determining the position of axes corners in ipython

* fixing axis position roundoff error locally

javadnoorb committed:
broken merge 

* minor change in comment

* added condition on inline backend

* added condition on inline backend

* issue with floats determining the position of axes corners in ipython

* fixing axis position roundoff error locally

* broken merge

javadnoorb committed:
Merge remote-tracking branch 'upstream/master'

javadnoorb committed:
Merge remote-tracking branch 'upstream/master'

javadnoorb committed:
update fork

Merge branch 'master' of https://github.com/pandas-dev/pandas

javadnoorb committed:
refactoring

javadnoorb committed:
refactoring

javadnoorb committed:
updating whatsnew docs

javadnoorb committed:
updating whatsnew docs

javadnoorb committed:
updating whatsnew docs

javadnoorb committed:
updating whatsnew docs

javadnoorb committed:
added more comments

javadnoorb committed:
Syncing

Merge remote-tracking branch 'upstream/master'

javadnoorb committed:
syncing

Merge branch 'master' of https://github.com/pandas-dev/pandas

javadnoorb committed:
fixing apply_frame_axis0

javadnoorb committed:
repushing

javadnoorb committed:
syncing the fork

Merge remote-tracking branch 'upstream/master'

javadnoorb commented:
Sorry for the git history mess

jbrockmendel commented:
Don’t worry about the git history.  We’ll definitely need a test though.

javadnoorb commented:
It passes pytest locally, but has failed travis/appveyor. As of the test it's not clear to me how to do it. The extra plot does not produce a handle (3 plots lead to 2 axes objects), so there's no straightforward way to count it. Basically the function  gets called, but no trace of it is in the output of `apply_frame_axis0`/`fast_apply`. Does using some sort of static flags and incrementing them in cython seem like a good idea? Or possibly a private flag in `fast_apply`? Any suggestions?

jbrockmendel commented:
> It passes pytest locally, but has failed travis/appveyor

Did you run the test suite in both py2 and py3?  Linux or OSX or Windows?

Sometimes there are Travis failures unrelated to the PR, which are often solved by just re-pushing to make Travis run again.

Plots in general are tough to test.  If you can get a test working using pytest-mpl that could be _really_ helpful.

javadnoorb commented:
It was in py3 OSX. I'm repushing to see if that works. 

I'll try to come up with a test for this in the next couple of days.

javadnoorb committed:
test added

javadnoorb commented:
I wrote a test that checks the number of runs for an arbitrary function (np.mean). It may be possible to write something similar for plots with class inheritance from pd.core.frame.DataFrame or instead use the suggested pytest-mpl. But given that this test already captures the issue in a more general sense, do we need an explicit test for scatter plots?

I'm still not sure why so many tests keep failing, but there seem to exist lines such as the following which failed by pytest:
https://github.com/pandas-dev/pandas/blob/de664e82086dc147ac63a09848f0b2f13befde5f/pandas/tests/groupby/test_groupby.py#L1398

I suspect there may be more cases like that. It seems to me that these need to be analyzed one by one and removed if the PR is expected go through. Should I go ahead and remove the ones that seem to be conflicting? I'm worried this may have too many unwanted consequences.

javadnoorb committed:
cleaner test

javadnoorb committed:
first group is already a copy

javadnoorb committed:
use demean in test

pep8speaks commented:
Hello @javadnoorb! Thanks for updating the PR.

 - In the file [`pandas/tests/groupby/test_groupby.py`](https://github.com/pandas-dev/pandas/blob/4d6402318fdad2cb0b0b5d092ae813b4f45c029a/pandas/tests/groupby/test_groupby.py), following are the PEP8 issues :

> [Line 37:1](https://github.com/pandas-dev/pandas/blob/4d6402318fdad2cb0b0b5d092ae813b4f45c029a/pandas/tests/groupby/test_groupby.py#L37): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

javadnoorb committed:
remove whitespace

toobaz commented on pandas/_libs/reduction.pyx:
```diff
@@ -518,6 +519,7 @@ def apply_frame_axis0(object frame, object f, object names,
         except:
             raise InvalidApply('Let this error raise above us')
 
+    piece = result
```
The ``try...except`` above is based on the assumption that ``result = f(chunk)`` can fail. I have no idea of when/why it does, but if it does, ``result`` is not defined and this line fails.

javadnoorb replied:
I assumed if exception is raised, program terminates and wouldn't reach this line. To be safe I'll move it inside the try clause, or maybe should rename result to piece

toobaz replied:
Sorry, you are absolutely right. What I should have written is: this will fail if ``n=0``. Then I have no idea if it can happen, but the code that follows doesn't fail.

toobaz commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -23,6 +23,38 @@
+    return arr - arr.mean()
+
+
+def test_apply_frame_axis0_runs_each_group_once(skip_first_index=True):
```
What is the role of a named parameter in a test?

toobaz replied:
(please add a reference to this bug)

toobaz commented on pandas/_libs/reduction.pyx:
```diff
@@ -494,7 +494,8 @@ class InvalidApply(Exception):
 def apply_frame_axis0(object frame, object f, object names,
-                      ndarray[int64_t] starts, ndarray[int64_t] ends):
+                      ndarray[int64_t] starts, ndarray[int64_t] ends,
+                      skip_first_index = True):
```
Would you please add a docstring, or at least clarify the role of ``skip_first_index``? (I can't get it)

javadnoorb replied:
sorry. `skip_first_index` was to do debugging quicker with cython. `skip_first_index = True` is with the changes. `skip_first_index = False` is without. I'll remove it and make a commit.

toobaz replied:
OK, no problem

jreback replied:
not sure why you need to add a parameter here? fast_apply (which calls this) is the first group operation, so by definition this is never called twice. The following code in pandas/core/groupby/ops.py *can* call the groups again, but that is only if the fast path has failed.

toobaz commented on pandas/_libs/reduction.pyx:
```diff
@@ -529,13 +531,14 @@ def apply_frame_axis0(object frame, object f, object names,
 
             object.__setattr__(slider.dummy, 'name', names[i])
-            piece = f(slider.dummy)
+            if i >= int(skip_first_index):
```
I would add a comment (here or above) about what you are doing, and a reference to the issue.

toobaz commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -23,6 +23,38 @@
+
+        def f(self, x):
+            self.counter += 1
+            return demean(x)
```
What about

```
from itertools import count
f = count()
...
assert f() = 3
```
?

javadnoorb replied:
Good idea. I'm not familiar with using count in this way without `__next__` method. Could you explain a bit more?

I tried the following, and it works:

```python
    from itertools import count
    cc = count()
    f = lambda x: cc.__next__()  # need a function with at least one input, otherwise there will be an error
    ...
    assert f(None) == 2
```

toobaz replied:
Sorry, my mistake again.

I think your solution only works with python 3.*, while
```
f = lambda x : next(cc)
```
should also work in 2.7. And yes, it will be ``==2``, not ``==3``. Maybe ``assert next(cc) == 2`` is slightly cleaner though.

TomAugspurger replied:
In the past, pytest's assertion rewriting had trouble with statements that have side-effects like this. I'd recommend

```python
result = next(cc)
assert result == 2
```

toobaz reviewed (changes requested):
Some comments. I think I liked the previous version better... maybe because I don't understand the use for ``skip_first_index=False``

javadnoorb commented:
@jreback, sorry, that parameter was just for debugging. I'll remove in the next commit. Some tests related in tests/groupby fail. I'll make a commit as soon as I find out the reason. 

Regarding your comment, I suspect you're referring to this line:
https://github.com/pandas-dev/pandas/blob/1f6ddc4592728c1a09fc0a81ca2a00c685b60a8e/pandas/core/groupby/ops.py#L181

But the two calls to f(group #0) don't happen in fast_apply. They happen inside apply_frame_axis0:
First here:
https://github.com/pandas-dev/pandas/blob/1f6ddc4592728c1a09fc0a81ca2a00c685b60a8e/pandas/_libs/reduction.pyx#L515
and then here:
https://github.com/pandas-dev/pandas/blob/1f6ddc4592728c1a09fc0a81ca2a00c685b60a8e/pandas/_libs/reduction.pyx#L532

javadnoorb commented:
I think PR name was confusing. I renamed it.

javadnoorb committed:
safe skipping

javadnoorb committed:
match style

javadnoorb committed:
match style

javadnoorb commented on pandas/_libs/reduction.pyx:
```diff
@@ -529,16 +529,20 @@ def apply_frame_axis0(object frame, object f, object names,
+            # issue #21609. Skipping the first group to avoid running
+            # the functions twice. For plots this leads to an extra
+            # plot of the first group.
+            if (i > 0) | (hasattr(piece, 'index')):
```
Catching mutation status of the first group turned out to be a headache. The way I have written this right now it only avoids double-runs of first group if function `f` does not output a dataframes/etc (which have index attribute), which is the case for plots.

javadnoorb commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1385,7 +1408,7 @@ def foo(x):
 
 def test_group_name_available_in_inference_pass():
-    # gh-15062
+    # gh-15062, issue #21609
```
If all double-runs (and not just the ones with output `index` attribute) were fixed, this test wouldn't have passed. So I included the issue number.

javadnoorb commented on pandas/_libs/reduction.pyx:
```diff
@@ -529,16 +529,20 @@ def apply_frame_axis0(object frame, object f, object names,
-            try:
+
+            if hasattr(piece, 'index'):
                 if piece.index is slider.dummy.index:
```
Changed `try/except` to more explicit `if`. No difference otherwise.

javadnoorb committed:
pytest-mpl added

javadnoorb commented:
The pytest-mpl test works locally (on a separate script file). How/where should I add it? Where does the baseline file go?

jbrockmendel commented:
I may have jumped the gun by suggesting pytest-mpl.  If you have another way to test the relevant behavior, go for it.

javadnoorb commented:
Ok, I'm gonna comment out the pytest-mpl for now to make sure the other tests pass, and wait for feedback on it. Catching the extra graph is not easy without manipulating plot functions (adding counters) so I actually don't think the test adds much beyond the one I've already written in tests/groupby.

javadnoorb committed:
pytest-mpl commented out temporarily

javadnoorb commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -11,6 +11,40 @@
+#         sdata, f, names, starts, ends)
+#
+#     return results[0]
+
```
Commenting this out until we figure if it's needed and/or where to put it.

javadnoorb commented:
For the record, after the changes, running the following in Jupyter lab:

```python
%matplotlib inline
import pandas as pd
df = pd.DataFrame([[1, 2], [3, 4], [5, 6], [7, 8]], columns=['x', 'y'])
df['cat'] = [1, 1, 1, 1]
df.groupby('cat').plot.scatter(x='x', y='y')
```

I get:

![image](https://user-images.githubusercontent.com/26352146/43092541-8e6f4c62-8e7b-11e8-89fc-6991e264e50e.png)

without the changes the output is:

![image](https://user-images.githubusercontent.com/26352146/43092900-96f75ba8-8e7c-11e8-81b2-b15d1220c99e.png)

javadnoorb committed:
new style class

javadnoorb committed:
re-push

javadnoorb committed:
re-push!!!

javadnoorb committed:
re-push!!!!

javadnoorb commented:
One Appveyor job keeps getting canceled  without a log. I see other tasks there with similar status. Does anyone know if this is a server issue or from my code?

javadnoorb committed:
trying np.int64 for windows

TomAugspurger commented:
I don't think we need pytest-mpl.

I'm a bit unsure about the approach here. I'd prefer to see a proper API for groupby that avoids the need to infer the output type, which the groupby plots can use.

javadnoorb commented:
@TomAugspurger Do you mean to skip `apply_frame_axis0 ` (or `fast_apply`) and instead take the slow path and loop through the groups when plots are passed as function?

TomAugspurger commented:
I'm not sure exactly what the API would be, but the goal is to give callers the ability to disable the "apply called twice" behavior. But this is a larger issue.

Since these groupby methods are internal, we have the ability to hardcode which functions aren't called twice. We do this already for things like `.hist`, `.boxplot`.

The following diff shows an example.

```diff
diff --git a/pandas/core/groupby/base.py b/pandas/core/groupby/base.py
index b2c5a8cff..65e6bdf35 100644
--- a/pandas/core/groupby/base.py
+++ b/pandas/core/groupby/base.py
@@ -59,6 +59,7 @@ class GroupByMixin(object):
 # special case to prevent duplicate plots when catching exceptions when
 # forwarding methods from NDFrames
 plotting_methods = frozenset(['plot', 'boxplot', 'hist'])
+extra_plotting_methods = frozenset(['scatter'])
 
 common_apply_whitelist = frozenset([
     'last', 'first',
diff --git a/pandas/core/groupby/groupby.py b/pandas/core/groupby/groupby.py
index 4b0143b3e..de59673dc 100644
--- a/pandas/core/groupby/groupby.py
+++ b/pandas/core/groupby/groupby.py
@@ -311,6 +311,7 @@ class GroupByPlot(PandasObject):
         def attr(*args, **kwargs):
             def f(self):
                 return getattr(self.plot, name)(*args, **kwargs)
+            f.__name__ = name
             return self._groupby.apply(f)
         return attr
 
diff --git a/pandas/core/groupby/ops.py b/pandas/core/groupby/ops.py
index f2c55a56b..1ba80cbf9 100644
--- a/pandas/core/groupby/ops.py
+++ b/pandas/core/groupby/ops.py
@@ -177,6 +177,7 @@ class BaseGrouper(object):
         # oh boy
         f_name = com._get_callable_name(f)
         if (f_name not in base.plotting_methods and
+                f_name not in base.extra_plotting_methods and
                 hasattr(splitter, 'fast_apply') and axis == 0):
             try:
                 values, mutated = splitter.fast_apply(f, group_keys)
```

Not claiming that's the perfect solution, but I think it's a bit cleaner than changing `apply_frame_axis0`

javadnoorb commented:
Thanks, didn't realize this was already there

javadnoorb commented:
@TomAugspurger Since you already have this written up do you want to push your version?

TomAugspurger committed:
Push diff

TomAugspurger commented:
FYI, you can apply diffs with `git apply`

```python
git apply foo.patch
```

You'll want to include other plot methods than scatter, revert the changes to `apply_frame_axis_0`, update the tests, etc.

javadnoorb commented:
Ok, thanks.

javadnoorb committed:
applying patch/reversion

javadnoorb commented:
@TomAugspurger your patch (plus added plot names) fixes all plot functions, when I try:
```python
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
df = pd.DataFrame([[1, 2], [3, 4], [5, 6], [7, 8]], columns=['x', 'y'])
df['cat'] = [1, 1, 1, 1]

extra_plotting_methods = ['line', 'bar', 'barh', 'box', 'kde', 'density', 'area', 'pie', 'scatter', 'hexbin']
p = df.groupby('cat').plot

for plotting_method in extra_plotting_methods:
    ax = p.__getattr__(plotting_method)(x='x', y='y');
    ax.iloc[0].set_title(plotting_method)
```

Still not sure how to implement a test though.

javadnoorb committed:
lint fix

javadnoorb committed:
test

javadnoorb committed:
setting next(..) to variable

javadnoorb committed:
pytest marks

javadnoorb committed:
pytest marks

javadnoorb committed:
CLN/TST: indents

javadnoorb committed:
TST: check scipy import

javadnoorb committed:
TST:skip_if_no_scipy

javadnoorb committed:
TST:skip_if_no_scipy

javadnoorb committed:
TST:skip_if_no_scipy for density

javadnoorb committed:
CLN/TST:lint

javadnoorb commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -11,6 +11,47 @@
+
+                def f(self):
+                    getattr(self.plot, name)(*args, **kwargs)
+                    # I have no idea why this is needed twice
```
RESOLVED: Function was returning next(counter_obj) when I wrote it.

javadnoorb committed:
TST: removed extra count

javadnoorb committed:
updating

TomAugspurger commented:
@javadnoorb looks like there are unrelated commits on this PR now, do you need help with git stuff? https://pandas.pydata.org/pandas-docs/stable/contributing.html#working-with-the-code might be helpful as well.

javadnoorb committed:
Merge branch 'master' into fix_apply_frame_axis0

TomAugspurger commented on pandas/tests/plotting/test_groupby.py:
```diff
@@ -11,6 +11,45 @@
+     ])
+@td.skip_if_no_mpl
+def test_no_double_plot_for_first_group(plotting_method):
+    class extGroupByPlot(pd.core.groupby.groupby.GroupByPlot):
```
I think this test can be simplified.

How about

```python
df = pd.DataFrame({'cat': [1, 1, 2, 2], 'x': [1, 3, 5, 7], 'y': [2, 4, 6, 8]})
df.groupby('cat').plot.scatter(x='x', y='y')
assert plt.get_fignums() == [1, 2]
```

or something like that.

TomAugspurger replied:
You'd need to replace `.plot.scatter` with your `getattr(gr, plotting_method)(x='x', y='y')`, but the basic idea is to test the output, the number of produced figures, is the number of groups.

javadnoorb replied:
That looks much better. I didn't know about `get_fignums`.

javadnoorb replied:
Do you by any chance know how to reset this? It works for the first plot method, but then keep accumulating as `pytest.mark.parametrize` is looped through.

javadnoorb replied:
Run `plt.close()` twice?

TomAugspurger replied:
Each test should close all its state. `plt.close('all')` should do it.

javadnoorb commented:
@TomAugspurger yeah. I messed up while syncing my fork. There were too many conflicts after the merge and I manually removed mine against the master, and some newlines were left here and there. Was I supposed to do this manually?

javadnoorb committed:
TST/CLN: simplify test/clean git-merge mess

javadnoorb committed:
TST: xref added

javadnoorb committed:
TST: comment issue

javadnoorb committed:
TST: adding close all before (in case)

javadnoorb committed:
timedout - repush

javadnoorb commented:
@jreback @toobaz @TomAugspurger. Just leaving a note. This should be ready for review. -thanks

TomAugspurger reviewed (commented):
It looks like there are two uses of `pandas.core.groupby.base.plotting_methods`.

1. core/groupby/ops.py, which you've updated
2. core/rgoupby/groupby.py in `_make_wrapper`, which hasn't been updated.

Can you try just adding these new methods to `plotting_methods` and seeing if anything breaks? Ideally we would just have the one frozenset. I initially added `extra_plotting_methods` just to test things out.

javadnoorb committed:
fixed comment

javadnoorb commented:
@TomAugspurger, I don't think `plotting_methods` and `extra_plotting_methods` should be merged. It leads to some errors like:

`AttributeError: type object 'Series' has no attribute 'hexbin'`

`plotting_methods` affects:

https://github.com/pandas-dev/pandas/blob/26b3e7de56fdebaf1842b01d747a1a30a126c54a/pandas/core/groupby/base.py#L80

and 

https://github.com/pandas-dev/pandas/blob/26b3e7de56fdebaf1842b01d747a1a30a126c54a/pandas/core/groupby/base.py#L86

which are then used as methods for dataframes and series.

I think only functions used in this doc can be methods of `groupby.plot`:

http://pandas.pydata.org/pandas-docs/version/0.16.2/generated/pandas.core.groupby.DataFrameGroupBy.plot.html

`hist` is the only exception because it has dual functionality (both for series/dataframes and groups). But I think it should be sufficient to have it in `plotting_methods`.

Following the other point, adding `extra_plotting_methods` to core/goupby/groupby.py:`_make_wrapper` doesn't seem to do anything when I try it. These `curried_with_axis` functions are pretty old and have no issue number. I can't track down why they were written, but seems like they just explicitly add an axis to the plot if none is provided.

TomAugspurger commented:
Thanks for the clarification on merging those two sets. Is the make_wrapper code dead? Does removing it break any tests? I’ve never touched them.

________________________________
From: Javad Noorbakhsh <notifications@github.com>
Sent: Monday, July 30, 2018 2:25:56 PM
To: pandas-dev/pandas
Cc: Tom Augspurger; Mention
Subject: Re: [pandas-dev/pandas] apply_frame_axis0 applies function to first group twice (#21943)


@TomAugspurger<https://github.com/TomAugspurger>, I don't think plotting_methods and extra_plotting_methods should be merged. It leads to some errors like:

AttributeError: type object 'Series' has no attribute 'hexbin'

plotting_methods affects:

https://github.com/pandas-dev/pandas/blob/26b3e7de56fdebaf1842b01d747a1a30a126c54a/pandas/core/groupby/base.py#L80

and

https://github.com/pandas-dev/pandas/blob/26b3e7de56fdebaf1842b01d747a1a30a126c54a/pandas/core/groupby/base.py#L86

which are then used as methods for dataframes and series.

I think only functions used in this doc can be methods of groupby.plot:

http://pandas.pydata.org/pandas-docs/version/0.16.2/generated/pandas.core.groupby.DataFrameGroupBy.plot.html

hist is the only exception because it has dual functionality (both for series/dataframes and groups). But I think it should be sufficient to have it in plotting_methods.

Following the other point, adding extra_plotting_methods to core/goupby/groupby.py:_make_wrapper doesn't seem to do anything when I try it. These curried_with_axis functions are pretty old and have no issue number. I can't track down why they were written, but seems like they just explicitly add an axis to the plot if none is provided.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pandas-dev/pandas/pull/21943#issuecomment-408981161>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABQHIlKw6qDVSnqrNnc8dHZOliHPC3Cyks5uL13EgaJpZM4VSJaX>.

javadnoorb commented:
No, it's not dead code. Just that it doesn't run when `groupby.plot.apply` is called. Commenting it out causes some tests to fail:
```
___________________________________________________________________________________________________ TestDataFrameGroupByPlots.test_series_groupby_plotting_nominally_works ___________________________________________________________________________________________________

self = <pandas.tests.plotting.test_groupby.TestDataFrameGroupByPlots object at 0x1a1fb72278>

    def test_series_groupby_plotting_nominally_works(self):
        n = 10
        weight = Series(np.random.normal(166, 20, size=n))
        height = Series(np.random.normal(60, 10, size=n))
        with tm.RNGContext(42):
            gender = np.random.choice(['male', 'female'], size=n)
    
        weight.groupby(gender).plot()
        tm.close()
>       height.groupby(gender).hist()

pandas/tests/plotting/test_groupby.py:49: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pandas.core.groupby.generic.SeriesGroupBy object at 0x1a1fc10d30>, attr = 'hist'

    def __getattr__(self, attr):
        if attr in self._internal_names_set:
            return object.__getattribute__(self, attr)
        if attr in self.obj:
            return self[attr]
        # if hasattr(self.obj, attr):
        #     return self._make_wrapper(attr)
    
        raise AttributeError("%r object has no attribute %r" %
>                            (type(self).__name__, attr))
E       AttributeError: 'SeriesGroupBy' object has no attribute 'hist'

pandas/core/groupby/groupby.py:537: AttributeError
```

javadnoorb committed:
CLN: adding comments to test/cleaning whatsnew entry

javadnoorb committed:
CLN: repush

javadnoorb committed:
syncing

Merge branch 'master' of https://github.com/pandas-dev/pandas into fix_apply_frame_axis0_tmp

javadnoorb committed:
CLN: adding comments

javadnoorb committed:
empty commit: failed tests seem unrelated

jreback commented:
closing as stale. if you'd like to continue working, pls ping.
